### PR TITLE
Future improvement for coverage exclusion, fix some scan warnings

### PR DESCRIPTION
**Pull request by asears**
## PR Summary

This introduces a new property to the csharp project CodeCoverageConfiguration.ExcludePaths.  It can be used for further implementation of configuration property in the Powershell code to exclude paths from Code Coverage results.  

Example feature for implementation would  be a script in the folder is picked up for coverage, however it is a Pester test harness script or other Powershell script not under test and not requiring coverage.

A couple minor updates for excluding Write-Host check from the scanner, spelling in the PSScriptAnalyzer custom check, fixing null check and alias scan warning.

## PR Checklist

- [ ] PR has meaningful title
- [ ] Summary describes changes
- [ ] PR is ready to be merged
  - If not, use the arrow next to `Create Pull Request` to mark it as a draft. PR can be marked `Ready for review` when it's ready.
- [ ] Tests are added/update *(if required)*
- [ ] Documentation is updated/added *(if required)*

### CODE REVIEW

File: src/functions/Coverage.ps1 [hunk 1]
```diff
@@ -87,7 +88,7 @@ function Exit-CoverageAnalysis {
     # to only get those that are not $null
     # (like if we did $breakpoints | where {$_ -ne $null})
     # so DON'T change this.
-    $breakpoints = @($CommandCoverage.Breakpoint) -ne $null
+    $breakpoints = null -ne @($CommandCoverage.Breakpoint)
```
Comment (ArmaanMcleod) [inline src/functions/Coverage.ps1:91]: null should be $null

File: src/csharp/Pester/Coverage.cs [hunk 2]
```diff
@@ -23,6 +23,7 @@ public static CodeCoverage Create()
         public List<object> CommandsMissed = new List<object>();
         public List<object> CommandsExecuted = new List<object>();
         public List<object> FilesAnalyzed = new List<object>();
+        public List<object> FIlesExcluded = new List<object>();
```
Comment (ArmaanMcleod) [inline src/csharp/Pester/Coverage.cs:26]:
```suggestion
        public List<object> FilesExcluded = new List<object>();
```

File: src/functions/Coverage.ps1 (hunk 1 above)
Comment (ArmaanMcleod) [inline src/functions/Coverage.ps1:91]:
```suggestion
    $breakpoints = $null -ne @($CommandCoverage.Breakpoint)
```

File: src/csharp/Pester/CodeCoverageConfiguration.cs [hunk 3]
```diff
@@ -46,6 +47,7 @@ public CodeCoverageConfiguration() : base("CodeCoverage configuration.")
             OutputPath = new StringOption("Path relative to the current directory where code coverage report is saved.", "coverage.xml");
             OutputEncoding = new StringOption("Encoding of the output file.", "UTF8");
             Path = new StringArrayOption("Directories or files to be used for codecoverage, by default the Path(s) from general settings are used, unless overridden here.", new string[0]);
+            ExcludePaths = new StringArrayOption("Directories or files to be excluded from codecoverage.", new string[0]);
```
Comment (ArmaanMcleod) [inline src/csharp/Pester/CodeCoverageConfiguration.cs:50]: Might be good to mention what default behavior is, just like Path option.

File: PSScriptAnalyzerSettings.psd1 [hunk 4]
```diff
@@ -6,5 +6,6 @@
         'PSUseShouldProcessForStateChangingFunctions'
         'PSUseApprovedVerbs'
         'Measure-SafeCommands'
+        'PSAvoidUsingWriteHost'
```
Comment (ArmaanMcleod) [inline PSScriptAnalyzerSettings.psd1:9]: Wouldn't this raise a lot of PSScriptAnalyzer warnings all over the codebase? We couldn't simply replace `Write-Host` with `Write-Output` given there is no `-BackgroundColor/-ForegroundColor/-NoNewLine` options for that cmdlet. Probably best doing this work in another issue anyways, doesn't seem to belong here in my opinion.

Comment (fflaten) [reply PSScriptAnalyzerSettings.psd1:9]: I guess that's why it's set under excludedrules. 🙂

**Comment by ArmaanMcleod** [comment]
Added some comments. I can probably help with the powershell implementation if you want. 

From first glance I think we will need to filter the files in https://github.com/pester/Pester/blob/main/src/functions/Coverage.ps1#L206.

**Comment by ArmaanMcleod** [comment]
Added some comments. I can probably help with the powershell implementation if you want. 

From first glance I think we will need to filter the files in https://github.com/pester/Pester/blob/main/src/functions/Coverage.ps1#L206.

File: PSScriptAnalyzerSettings.psd1 (hunk 4 above)
Comment (ArmaanMcleod) [reply PSScriptAnalyzerSettings.psd1:9]: Oops, I guess thats what I get for looking at code late in the night 😄

Comment (asears) [reply PSScriptAnalyzerSettings.psd1:9]: It would be great if someone could put in a PR to turn off that rule in PSScriptAnalyzer by default.

**Comment by asears** [comment]
@ArmaanMcleod I would appreciate someone taking the powershell implementation on after this PR gets merged.

**Comment by asears** [comment]
Not clear why the tests are failing with this PR...

File: src/functions/Coverage.ps1 (hunk 1 above)
Comment (fflaten) [reply src/functions/Coverage.ps1]: Did you read the comment above this line?

**Comment by ArmaanMcleod** [comment]
> @ArmaanMcleod I would appreciate someone taking the PowerShell implementation on after this PR gets merged.

To me it makes sense to have the C# and PowerShell code changes in this PR, since then you can add tests to validate the option is working correctly. @nohwnd What do you think?

**Pull request closed without merging**